receiver/cudareceiver/collector.go

```diff
@@ -0,0 +1,179 @@
+
+	startTime time.Time
+	device    *NVMLDevice
+
+	scrapeInterval time.Duration
+	metricPrefix   string
+	done           chan struct{}
+}
+
+// NewCUDAMetricsCollector creates a new set of CUDA (GPU) Metrics (temperature, power, et al.).
+func NewCUDAMetricsCollector(d time.Duration, prefix string, consumer consumer.MetricsConsumerOld) (*CUDAMetricsCollector, error) {
+	device, status := NVMLDeviceGetHandledByIndex(uint64(0))
+	if status != NVMLSuccess {
+		return nil, fmt.Errorf("Could not get GPU device: status=%d", status)
+	}
+
+	c := &CUDAMetricsCollector{
+		consumer:       consumer,
+		startTime:      time.Now(),
+		device:         device,
+		scrapeInterval: d,
+		metricPrefix:   prefix,
+		done:           make(chan struct{}),
+	}
+
+	return c, nil
+}
+
+func detectResource() {
+	resourceDetectionSync.Do(func() {
+		res, err := auto.Detect(context.Background())
+		if err != nil {
+			panic(fmt.Sprintf("Resource detection failed, err:%v", err))
+		}
+		if res != nil {
+			rsc = &resourcepb.Resource{
+				Type:   res.Type,
+				Labels: make(map[string]string, len(res.Labels)),
+			}
+			for k, v := range res.Labels {
+				rsc.Labels[k] = v
+			}
+		}
+	})
+}
+
+// StartCollection starts a ticker'd goroutine that will scrape and export CUDA metrics periodically.
+func (c *CUDAMetricsCollector) StartCollection() {
+	detectResource()
+	status := NVMLInit()
+	if status != NVMLSuccess {
+		// TODO: handle inappropriate GPU state
+		return
+	}
+
+	go func() {
+		ticker := time.NewTicker(c.scrapeInterval)
+		for {
+			select {
+			case <-ticker.C:
+				c.scrapeAndExport()
+
+			case <-c.done:
+				return
+			}
+		}
+	}()
+}
+
+// StopCollection stops the collection of metric information
+func (c *CUDAMetricsCollector) StopCollection() {
+	status := NVMLShutdown()
+	if status != NVMLSuccess {
+		// TODO: handle inappropriate GPU state
+		return
+	}
+	close(c.done)
+}
+
+// scrapeAndExport
+func (c *CUDAMetricsCollector) scrapeAndExport() {
+	ctx := context.Background()
+
+	metrics := make([]*metricspb.Metric, 0, len(cudaMetricDescriptors))
+	var errs []error
+
+	metrics = append(
+		metrics,
+		&metricspb.Metric{
+			MetricDescriptor: metricTemperature,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.Temperature())},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricPower,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.PowerUsage())},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricPCIeThroughputTX,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.PCIeThroughput(PCIeUtilTXBytes))},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricPCIeThroughputRX,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.PCIeThroughput(PCIeUtilRXBytes))},
+		},
+	)
+
+	if len(errs) > 0 {
+		// TODO: emit error log
```

Comment:
Can we emit it now instead of leaving a TODO?

Reply:
Done